Cabal/tests/PackageTests/PackageTester.hs

```diff
@@ -1,367 +1,540 @@
-unregister config libraryName = do
-    res@(_, _, output) <- run Nothing (ghcPkgPath config) []
-                          ["unregister", "--user", libraryName]
-    if "cannot find package" `isInfixOf` output
-        then return ()
-        else requireSuccess res
-
--- | Install this library in the user area
-cabal_install :: SuiteConfig -> PackageSpec -> IO Result
-cabal_install config spec = do
-    buildResult <- doCabalBuild config spec
-    res <- if successful buildResult
-        then do
-            res <- cabal config spec [] ["install"]
-            return $ recordRun res InstallSuccess buildResult
-        else
-            return buildResult
-    record spec res
-    return res
-
-cabal_register :: SuiteConfig -> PackageSpec -> [String] -> IO Result
-cabal_register config spec extraArgs = do
-    res <- doCabalRegister config spec extraArgs
-    record spec res
-    return res
-
-doCabalRegister :: SuiteConfig -> PackageSpec -> [String] -> IO Result
-doCabalRegister config spec extraArgs = do
-    configResult <- doCabalConfigure config spec
-    if successful configResult
-        then do
-            buildResult <- doCabalBuild config spec
-            if successful buildResult
-              then do res <- cabal config spec [] ("register" : extraArgs)
-                      return $ recordRun res RegisterSuccess configResult
-              else return buildResult
-        else
-            return configResult
-
-
-cabal_test :: SuiteConfig -> PackageSpec -> [(String, Maybe String)]
-           -> [String] -> IO Result
-cabal_test config spec envOverrides extraArgs = do
-    res <- cabal config spec envOverrides ("test" : extraArgs)
-    let r = recordRun res TestSuccess nullResult
-    record spec r
-    return r
+------------------------------------------------------------------------
+-- * Compiling setup scripts
+
+compileSetup :: TestM ()
+compileSetup = do
+    (suite, test) <- ask
+    pkg_path <- packageDir
+    liftIO $ rawCompileSetup (suiteVerbosity suite) suite (testEnvironment test) pkg_path
+
+rawCompileSetup :: Verbosity -> SuiteConfig -> [(String, Maybe String)] -> FilePath -> IO ()
+rawCompileSetup verbosity suite e path = do
+    r <- rawRun verbosity (Just path) (ghcPath suite) e $
+        [ "--make"] ++
+        ghcPackageDBParams (packageDBStack suite) ++
+        [ "-hide-all-packages"
+        , "-package base"
+        -- This is better, but we don't necessarily have it
+        -- if we're bootstrapping with old Cabal.
+        -- , "-package-id " ++ CURRENT_PACKAGE_KEY
+        , "-package Cabal"
+        -- GHC doesn't seem to necessarily reliably determine
+        -- when to rebuild Setup, so this is safest.
+        , "-fforce-recomp"
+        , "-O0"
+        , "Setup.hs" ]
+    unless (resultExitCode r == ExitSuccess) $
+        error $
+        "could not build shared Setup executable\n" ++
+        "  ran: " ++ resultCommand r ++ "\n" ++
+        "  output:\n" ++ resultOutput r
```

Comment:
This is not a new thing, but I think readability could be improved by adding a few more newlines (in the output!) around the resultOutput bit